dc/s2n-quic-dc/src/socket/recv/descriptor.rs

```diff
@@ -0,0 +1,446 @@
+
+        // based on the implementation in:
+        // https://github.com/rust-lang/rust/blob/28b83ee59698ae069f5355b8e03f976406f410f5/library/alloc/src/sync.rs#L2551
+        if desc_ref != 1 {
+            trace!(drop_desc_ref = inner.id);
+            return;
+        }
+
+        core::sync::atomic::fence(Ordering::Acquire);
+
+        let mem = inner.free(self);
+
+        trace!(free_desc = inner.id, state = %"filled");
+
+        drop(mem);
+    }
+
+    #[inline]
+    pub(super) fn drop_unfilled(&self) {
+        let inner = self.inner();
+        inner.references.store(0, Ordering::Release);
+        let mem = inner.free(self);
+
+        trace!(free_desc = inner.id, state = %"unfilled");
+
+        drop(mem);
+    }
+}
+
+unsafe impl Send for Descriptor {}
+unsafe impl Sync for Descriptor {}
+
+pub(super) struct DescriptorInner {
+    id: u64,
+    address: NonNull<Addr>,
+    data: NonNull<u8>,
+
+    references: AtomicUsize,
+
+    memory: NonNull<Memory>,
+}
+
+impl DescriptorInner {
+    pub(super) fn new(
+        id: u64,
+        address: NonNull<Addr>,
+        data: NonNull<u8>,
+        memory: NonNull<Memory>,
+    ) -> Self {
+        Self {
+            id,
+            address,
+            data,
+            references: AtomicUsize::new(0),
+            memory,
+        }
+    }
+
+    #[inline]
+    fn capacity(&self) -> u16 {
+        unsafe { self.memory.as_ref().capacity }
+    }
+
+    /// Frees the descriptor back into the pool
+    #[inline]
+    fn free(&self, desc: &Descriptor) -> Option<Box<Memory>> {
+        let memory = unsafe { self.memory.as_ref() };
+        let mem_refs = memory.references.fetch_sub(1, Ordering::Release);
+        debug_assert_ne!(mem_refs, 0, "reference count underflow");
+
+        // if the free_list is still active (the allocator hasn't dropped) then just push the id
+        // TODO Weak::upgrade is a bit expensive since it clones the `Arc`, only to drop it again
+        if let Some(free_list) = memory.free_list.upgrade() {
+            free_list.free(Descriptor {
+                ptr: desc.ptr,
+                phantom: PhantomData,
+            });
+            return None;
+        }
+
+        // the free_list no longer active and we need to clean up the memory
+
+        // based on the implementation in:
+        // https://github.com/rust-lang/rust/blob/28b83ee59698ae069f5355b8e03f976406f410f5/library/alloc/src/sync.rs#L2551
+        if mem_refs != 1 {
+            trace!(memory_draining = mem_refs - 1, desc = self.id);
+            return None;
+        }
+
+        core::sync::atomic::fence(Ordering::Acquire);
+
+        trace!(memory_free = ?self.memory.as_ptr(), desc = self.id);
+
+        // return the boxed memory rather than free it here - this works around
+        // any stacked borrowing issues found by Miri
+        Some(unsafe { Box::from_raw(self.memory.as_ptr()) })
+    }
+}
+
+/// An unfilled packet
+pub struct Unfilled {
+    desc: Option<Descriptor>,
```

Comment:
Just to check, this Option is just because you can't move out of the field given the Drop impl?

Reply:
Correct. And the `impl Drop` needs to know if the descriptor was consumed in `recv_with` or was dropped by the owner.